### PR TITLE
[Rust]: Refactor the existing code and add new APIs in wasmedge-sys

### PR DESCRIPTION
**Pull request by apepkuss**
### Motivation
This PR is to present some improvements on the existing object definitions and the relevant APIs in wasmedge-sys, and also to introduce some new objects and APIs based on [wasmedge.h.](https://github.com/WasmEdge/WasmEdge/blob/master/include/api/wasmedge/wasmedge.h).

### Details
The following iterms are involved:
- Refactored items
  - Config
  - ImportObject
    - [x] #746 
    - [x] #747 
    - [x] #749 

  - Function
  - Table
    - [x] #797 

  - Memory
    - [x] #767 

  - Global
    - [x] #733 
    - [x] #736 
    - [x] #741 

  - Module
  - Vm
    - [x] #770 
    - [x] #771 
    - [x] #772 
    - [x] #779 

  - error.rs

- New Items
  - Executor
    - [x] #700 
    - [x] #701 
    - [x] #702 
    - [x] #703 
    - [x] #704 
    - [x] #705 
    - [x] #709

  - Statistics
    - [x] #790 
    - [x] #791 
    - [x] #792 
    - [x] #793 
    - [x] #794 
    - [x] #840 
    - [x] #841 

  - Store
    - [x] #796
    - [x] #813 
    - [x] #814
    - [x] #815
    - [x] #816 
    - [x] #817 
    - [x] #818 
    - [x] #819 
    - [x] #820 
    - [x] #821 
    - [x] #822 
    - [x] #823 
    - [x] #824 
    - [x] #825 
    - [x] #826 
    - [x] #827 
    - [x] #828 
    - [x] #829 
    - [x] #830 
    - [x] #831 
    - [x] #832 
    - [x] #833 
    - [x] #834 
    - [x] #835 
    - [x] #836 
    - [x] #837 
    - [x] #838 
    - [x] #839 

  - types.rs
    - MemType
      - [x] #786 
      - [x] #787
      - [x] #789  

  - utils.rs

### Other Issues
This PR also relates to the following issues:
- #601 
- #635 
- #628

### CODE REVIEW

**Comment by yanganto** [comment]
Hi @apepkuss, 
There are a lot of things more than Rust SDK in the repo, please use `[Rust]` as prefix in you commit.  
Thanks.

**Comment by apepkuss** [comment]
Hi @yanganto,
Really thanks for your reminder, and I'll update the commit message to follow the convention. Thanks again!:D

**Comment by hydai** [comment]
Hi @apepkuss 
Looks like your history contains several commits which doesn't belong to your PR.

截圖 2021-11-30 at 13 23 39@2x

For the development, please use `git fetch/git rebase` instead of any `git pull/git merge` to make sure your commit history is a linear history.

**Comment by hydai** [comment]
Also we require every developer signing the DCO. For details, please refer to this doc: https://www.secondstate.io/articles/dco/

**Comment by yanganto** [comment]
Hi @apepkuss, may you clean your PR? 
There are a lot of commits already in master. 

You can try to rebase your branch on top of the master. 
If there are issues to do so, you can open a new branch on top of the master branch, then pick the commits you want, and open a PR again.

Thanks.

**Comment by apepkuss** [comment]
> Hi @apepkuss Looks like your history contains several commits which doesn't belong to your PR.
> 
> 截圖 2021-11-30 at 13 23 39@2x
> 
> For the development, please use `git fetch/git rebase` instead of any `git pull/git merge` to make sure your commit history is a linear history.

Hi @hydai, thanks for your reminder. I conducted the rebase before, which introduced the commits you mentioned. I will open a new PR to solve the issues. Thanks again!

**Comment by apepkuss** [comment]
> 

Hi @yanganto, thanks for your comments and suggestions. I'll open a new PR to solve the issue. Thanks again!

File: bindings/rust/wasmedge-sys/src/import_obj.rs [hunk 1]
```diff
@@ -1,67 +1,129 @@
...
+        }
+        global.registered = true;
+        global.ctx = std::ptr::null_mut();
+    }
+
+    pub fn init_wasi<T: Iterator<Item = E>, E: AsRef<str>>(
+        &self,
+        args: Option<T>,
+        envs: Option<T>,
+        dirs: Option<T>,
```
Comment (ZhangHanDong) [inline bindings/rust/wasmedge-sys/src/import_obj.rs:71]: Wasn't this `dir` already removed from wasmedge-sdk?

File: bindings/rust/wasmedge-sys/src/module.rs [hunk 2]
```diff
@@ -1,33 +1,119 @@
...
+
+    pub fn load_from_buffer(
         config: &crate::config::Config,
-        path: CString,
+        buffer: Vec<u8>,
     ) -> Result<Self, ErrReport> {
         let loader_ctx = unsafe { wasmedge::WasmEdge_LoaderCreate(config.ctx) };
         let mut ctx: *mut wasmedge::WasmEdge_ASTModuleContext = std::ptr::null_mut();
 
+        assert!(!buffer.is_empty(), "WasmEdge fail to load an empty buffer");
```
Comment (ZhangHanDong) [inline bindings/rust/wasmedge-sys/src/module.rs:50]: nice! This assertion is guaranteed to be safety. 
This unsafe function should be marked with a document comment to indicate that it is safe, as is the usual practice. 
```rust
/// # Safety 
///  Avoid the problem of uninitialized memory due to the introduction of an empty buffer
```

File: bindings/rust/wasmedge-sys/src/import_obj.rs (hunk 1 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/import_obj.rs:71]: Thanks for your comments. This PR is replaced by #854, please add your comments there. Thanks again!

File: bindings/rust/wasmedge-sys/src/module.rs (hunk 2 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/module.rs:50]: Thanks for your comments. This PR is replaced by #854, please add your comments there. Thanks again!

**Comment by apepkuss** [comment]
This PR is closed and replaced by the new PR #854.

**Pull request closed without merging**